### PR TITLE
Sprocket event marshal [CU-3u734kc]

### PR DESCRIPTION
**Pull request by bwpikaard**

### CODE REVIEW

**Comment by 0c370t** [comment]
Task linked: [CU-3u734kc Notification Service Decorator Based Events](https://app.clickup.com/t/3u734kc)

File: common/src/events/marshal/marshal.types.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import {z} from "zod";
+
+import type {EventPayload} from "../events.types";
+import {EventTopic} from "../events.types";
+
+export type EventFunction<Event extends EventTopic> = (data: EventPayload<Event>) => Promise<void>;
```
Comment (zachstence) [inline common/src/events/marshal/marshal.types.ts:6]: Not all event functions will be `async`

```suggestion
export type EventFunction<Event extends EventTopic> = (data: EventPayload<Event>) => void | Promise<void>;
```

**Review by zachstence — CHANGES_REQUESTED** [review]
This is awesome, can't wait to use this more throughout the code

🚀

File: common/src/events/marshal/marshal.ts [hunk 2]
```diff
@@ -39,7 +39,13 @@ export abstract class SprocketEventMarshal {
                         return;
                     }
 
-                    eventFunction(payload.data).catch(e => { this.logger.error(e) });
+                    try {
+                        // No need to catch; it's in a try-catch block because we don't know if it's a promise or not.
+                        // eslint-disable-next-line @typescript-eslint/no-floating-promises
+                        eventFunction(payload.data);
```
Comment (zachstence) [inline common/src/events/marshal/marshal.ts:45]: If its async, you have to `await` the Promise in order for the error to be caught in the try/catch.

You should be able to use `instanceof` to determine if its a Promise or not

```ts
try {
    const maybePromise = someFunction();
    if (maybePromise instanceof Promise) {
        await maybePromise;
    }
} catch (e) {
    // ...
}
```

**Review by zachstence — CHANGES_REQUESTED** [review]

**Review by zachstence — APPROVED** [review]

**Pull request merged**